OCM-3006 | feat: Add update functionality to HTPasswd IDPs

hunterkepley opened the pull request:
<!--
- Please ensure code changes are split into a series of logically independent commits.
- Every commit should have a subject/title (What) and a description/body (Why).
  The commit subject needs to conform to the following format:
  [JIRA-TICKET] | [TYPE]: <MESSAGE>
  Where:
     JIRA_TICKET is jira ticket ID (for example OCM-xxx)
     TYPE must be one of the options (case sensitive):
          feat, fix, docs, style, refactor, test, chore, build, ci, perf
  For more info see [Commit Messages Format Guide](../CONTRIBUTE.md)
- Every PR must have a description.
- As an example you can use git commit -m"What" -m"Why" to achieve the requirements above. GitHub automatically recognises the commit description (-m"Why") in single commit PRs and adds it as the PR description.
-->

**What this PR does / why we need it**:

Adds functionality to HTPasswd IDPs allowing users to be added/updated/deleted

**Which issue(s) this PR fixes** *(optional, use `fixes #<issue_number>(, fixes #<issue_number>, ...)` format, where issue_number might be a GitHub issue, or a Jira story (OCM-xxxx)*:
Fixes #[OCM-3006](https://issues.redhat.com//browse/OCM-3006)

**Change type**
- [X] New feature
- [ ] Bug fix
- [ ] Build
- [ ] CI
- [ ] Documentation
- [ ] Performance
- [ ] Refactor
- [ ] Style
- [ ] Unit tests
- [ ] Subsystem tests

https://issues.redhat.com/browse/OCM-3006

hunterkepley commented on examples/create_rosa_sts_cluster/classic_sts/cluster/output.tf:
```diff
@@ -11,10 +11,10 @@ output "oidc_endpoint_url" {
 }
 
 output "cluster_admin_username" {
-  value = rhcs_cluster_rosa_classic.rosa_sts_cluster.admin_credentials == null? null:rhcs_cluster_rosa_classic.rosa_sts_cluster.admin_credentials.username
```
This file's changes came from `make fmt`

hunterkepley commented on examples/create_identity_provider/htpasswd/README.md:
```diff
@@ -6,7 +6,7 @@ Using htpasswd authentication in OpenShift Container Platform allows you to iden
 
 ## Prerequisites
 
-1. You created your [account roles using Terraform](../../examples/create_rosa_cluster/create_rosa_sts_cluster/classic_sts/account_roles/README.md).
```
Updated this as i needed the link and it was incorrect :sweat_smile:

nirarg replied:
Thank you for that.
The examples here are going to be replaced soon with the examples in the module.
https://github.com/terraform-redhat/terraform-rhcs-rosa-classic 
Please check this from the module as well

gdbranco commented on provider/identityprovider/identity_provider_resource.go:
```diff
@@ -581,7 +584,114 @@ func (r *IdentityProviderResource) Read(ctx context.Context, request resource.Re
+
+		for user, htpasswdUser := range stateUserMap {
+			if planUserMap[user].Username == "" { // Not in plan, delete
+				err = htpasswd.DeleteUser(ctx, htpasswdUser.Id, resource)
```
I still think we need to setup a separate package for all of these, there's just a lot of things in here which are only for the htpsswd idp

gdbranco replied:
Oh you did it, but internally I see.

I wonder if we can capture the whole patch of htpsswd into a function then. Just because when we get to the other idps it will be a lot of code

gdbranco replied:
Just wrapping into a function call, like patchHtpasswordUserList or something, so from the update flow it is clearer what is happening. Eventually we'll have the other idps too, so code will get confusing

hunterkepley replied:
Tried moving all the logic to this pkg, including even moving the `htpasswd.go` file, but I cannot get around having an import cycle due to the state/plan vars. Any suggestions on this?
The API funcs are fine due to them not using the state nor plan, but moving the other logic over creates this cycle. The `IdenittyProviderState` struct is part of a file shared by all identity providers.

Technically, I could move the resource file into its own package, but im not sure if that's best.

hunterkepley replied:
i think tomorrow AM im gonna try to move all the files which import the resource file into their own dirs, not sure if thatll work but just thought of it now

nirarg replied:
@hunterkepley I don't understand the issue here
What prevents from moving all this code to dedicated update function inside the htpasswd file?
Everything added here from line 613 is specific for htpasswd, no?

hunterkepley replied:
@nirarg There is an import cycle preventing me from making the binary, it is due to the `IdentityProviderState` and resource structs. I can move the state file to its own package, but it's shared by all identityproviders. I will do this if you think that's alright, but i cannot move this code with the current file structure

nirarg commented on provider/identityprovider/htpasswd/api.go:
I think this file should be moved to the common repository
https://github.com/openshift-online/ocm-common
There is nothing here related to terraform and it can be used from the cli as well

hunterkepley replied:
I believe this is unrelated to this ticket, could we have another ticket for moving / editing the READMEs?

hunterkepley replied:
I think there is a lot to  refactor with IDPs, and we should tackle it in a cleanup epic/ticket

nirarg replied:
Please open a ticket including all details about the refactoring needed

hunterkepley replied:
Sure, will do at some point, want to make a meeting with @gdbranco and go over what all could be done as he had some ideas as well

sagidayan commented:
Can you please add tests?

hunterkepley commented:
> Can you please add tests?

@sagidayan Will add tests tomorrow, as it is day of learning

@nirarg I have moved a great deal of logic into the htpasswd package, leaving mostly function calls & error handling in the update function.

nirarg commented on provider/identityprovider/identity_provider_resource.go:
```diff
@@ -581,7 +583,76 @@ func (r *IdentityProviderResource) Read(ctx context.Context, request resource.Re
+		if err != nil {
+			response.Diagnostics.AddError(
+				"Can't get identity provider user ID",
+				fmt.Sprintf(
+					"Can't get identity provider user ID with identifier for "+
+						"cluster '%s': %v",
+					state.Cluster.ValueString(), err,
+				),
+			)
+			return
+		}
+		planUserMap, err := HTPasswdUserListToStringMaps(ctx, plan.HTPasswd.Users, resource)
+		if err != nil {
+			response.Diagnostics.AddError(
+				"Can't get identity provider user ID",
+				fmt.Sprintf(
+					"Can't get identity provider user ID with identifier for "+
+						"cluster '%s': %v",
+					state.Cluster.ValueString(), err,
+				),
+			)
+			return
+		}
+
+		patchParams := htpasswd.PatchParams{
+			ctx, stateUserMap, planUserMap, resource,
+			[]string{}, state.Cluster.ValueString(), response,
+		}
+
+		var ok bool
+		patchParams.RemovedUsers, ok = htpasswd.DeleteUserFromState(patchParams)
+		if !ok {
+			return
+		}
+
+		ok = htpasswd.PatchOrAddUserInState(patchParams)
+		if !ok {
+			return
+		}
+	}
```
What do you say about moving all this code to htpasswd.go as well?
This code is relted only to htpasswd
Call to function like:
```
updateHTPasswd(state.HTPasswd, plan.HTPasswd)
```

hunterkepley replied:
SGTM

nirarg commented on provider/identityprovider/htpasswd.go:
```diff
@@ -119,3 +120,27 @@ func uniqueUsernameValidator() validator.List {
 	})
 }
+
+func HTPasswdUserListToStringMaps(ctx context.Context, users []HTPasswdUser,
```
The fuction can be private as it is used only from within the package
```suggestion
func hTPasswdUserListToStringMaps(ctx context.Context, users []HTPasswdUser,
```

gdbranco commented on provider/identityprovider/htpasswd.go:
```diff
@@ -144,3 +146,49 @@ func HTPasswdUserListToStringMaps(ctx context.Context, users []HTPasswdUser,
+
+func UpdateHTPasswd(ctx context.Context, resource *cmv1.IdentityProviderClient, state *IdentityProviderState,
+	plan *IdentityProviderState, response *resource.UpdateResponse) {
+	if !reflect.DeepEqual(state.HTPasswd.Users, plan.HTPasswd.Users) {
```
Can we invert this condition? If it's equal return early

nirarg commented on provider/identityprovider/htpasswd.go:
```diff
@@ -119,3 +122,74 @@ func uniqueUsernameValidator() validator.List {
+	var ok bool
+	patchParams.RemovedUsers, ok = htpasswd.DeleteUserFromState(patchParams)
+	if !ok {
+		return
```
Shouldn't it return an error in case of !ok?

hunterkepley replied:
It already adds the error to the `Response.Diagnostics` - but i can move this to `htpasswd.go` instead

hunterkepley replied:
Only weird thing is that `Response.Diagnostics` takes a `string` not an `error`

nirarg commented:
/approve
Thank you @hunterkepley for adding this
Please make sure to fix all failures:
* Fix subsytem tests (run `make subsystem-test`)
* To fix the commit messages validation, I think that rebase the branch on `origin/main` will fix the issue
@gdbranco please add `lgtm` label once you think it ready
Is this planed to be released as part of v1.6.0? @ciaranRoche

hunterkepley commented:
> /approve Thank you @hunterkepley for adding this Please make sure to fix all failures:
> 
>     * Fix subsytem tests (run `make subsystem-test`)
> 
>     * To fix the commit messages validation, I think that rebase the branch on `origin/main` will fix the issue
>       @gdbranco please add `lgtm` label once you think it ready
>       Is this planed to be released as part of v1.6.0? @ciaranRoche

Fixed tests I think, but the commit msg validation is messed up or something, i have rebased (nothing changed, it's current) as well as verified the commit msg looks okay, but i get this error: `Invalid symmetric difference expression` @nirarg

hunterkepley commented:
/test

hunterkepley commented:
/test all

hunterkepley commented:
/test all

hunterkepley commented:
> Same as [#563 (comment)](https://github.com/terraform-redhat/terraform-provider-rhcs/pull/563#issuecomment-1980390316) We should sync with QE first to ensure they have the capacity and potentially test this branch before merging to gain confidence it does not impact 1.6.0

^ Do not merge until it is mentioned in the comments it is OK to do so

xueli181114 commented:
If our target date of 1.6.0 is end of March, QE has not enough capacity to finish the testing of this new feature before target date. It may cause high risk of bug leak.

hunterkepley commented:
@xueli181114 Hi :wave: - Is there an update on the release? Would like to get this merged :smile:

hunterkepley commented:
@nirarg @sagidayan PTAL, I believe the freeze has been lifted

gdbranco commented:
/ok-to-test

gdbranco commented:
/lgtm
